// @react-spring/zdog/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var reactZdog = require('react-zdog');
var core = require('@react-spring/core/index.cjs.js');
var stringInterpolation = require('@react-spring/shared/stringInterpolation');
var colorNames = _interopDefault(require('@react-spring/shared/colors'));
var animated$1 = require('@react-spring/animated/index.cjs.js');

var elements = ['Anchor', 'Shape', 'Group', 'Rect', 'RoundedRect', 'Ellipse', 'Polygon', 'Hemisphere', 'Cylinder', 'Cone', 'Box']; // Extend animated with all the available Zdog elements

var animated = animated$1.extendAnimated(animated$1.withAnimated, elements);

if (reactZdog.addEffect) reactZdog.addEffect(core.update);
core.Globals.assign({
  defaultElement: reactZdog.Anchor,
  frameLoop: reactZdog.addEffect && new core.FrameLoop({
    requestFrame: reactZdog.invalidate
  }),
  applyAnimatedValues: reactZdog.applyProps,
  createStringInterpolator: stringInterpolation.createStringInterpolator,
  colorNames: colorNames
});

Object.keys(core).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return core[k];
    }
  });
});
exports.a = animated;
exports.animated = animated;
