// @react-spring/addons/parallax.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutPropertiesLoose = _interopDefault(require('@babel/runtime/helpers/objectWithoutPropertiesLoose'));
var React = require('react');
var React__default = _interopDefault(React);
var globals = require('@react-spring/shared/globals');
var core = require('@react-spring/core/index.cjs.js');
var animated = require('@react-spring/animated/index.cjs.js');
var useMemoOne = require('use-memo-one');
var shared = require('@react-spring/shared');

var AnimatedView = animated.withAnimated(globals.defaultElement);
var ParentContext = React__default.createContext(null);

function getScrollType(horizontal) {
  return horizontal ? 'scrollLeft' : 'scrollTop';
}

var START_TRANSLATE_3D = 'translate3d(0px,0px,0px)';
var START_TRANSLATE = 'translate(0px,0px)';
var ParallaxLayer = React__default.memo(function (_ref) {
  var _extends2;

  var horizontal = _ref.horizontal,
      _ref$factor = _ref.factor,
      factor = _ref$factor === void 0 ? 1 : _ref$factor,
      _ref$offset = _ref.offset,
      offset = _ref$offset === void 0 ? 0 : _ref$offset,
      _ref$speed = _ref.speed,
      speed = _ref$speed === void 0 ? 0 : _ref$speed,
      rest = _objectWithoutPropertiesLoose(_ref, ["horizontal", "factor", "offset", "speed"]);

  // Our parent controls our height and position.
  var parent = React.useContext(ParentContext); // This is how we animate.

  var ctrl = useMemoOne.useMemoOne(function () {
    var targetScroll = Math.floor(offset) * parent.space;
    var distance = parent.space * offset + targetScroll * speed;
    return new core.Controller({
      space: parent.space * factor,
      translate: -(parent.current * speed) + distance
    });
  }, []); // Create the layer.

  var layer = useMemoOne.useMemoOne(function () {
    return {
      setPosition: function setPosition(height, scrollTop, immediate) {
        if (immediate === void 0) {
          immediate = false;
        }

        var targetScroll = Math.floor(offset) * height;
        var distance = height * offset + targetScroll * speed;
        ctrl.start({
          translate: -(scrollTop * speed) + distance,
          config: parent.config,
          immediate: immediate
        });
      },
      setHeight: function setHeight(height, immediate) {
        if (immediate === void 0) {
          immediate = false;
        }

        ctrl.start({
          space: height * factor,
          config: parent.config,
          immediate: immediate
        });
      }
    };
  }, []); // Register the layer with our parent.

  shared.useOnce(function () {
    if (parent) {
      parent.layers.add(layer);
      parent.update();
      return function () {
        parent.layers.delete(layer);
        parent.update();
      };
    }
  });
  var translate3d = ctrl.get('translate').to(horizontal ? function (x) {
    return "translate3d(" + x + "px,0,0)";
  } : function (y) {
    return "translate3d(0," + y + "px,0)";
  });
  return React__default.createElement(AnimatedView, Object.assign({}, rest, {
    style: _extends((_extends2 = {
      position: 'absolute',
      backgroundSize: 'auto',
      backgroundRepeat: 'no-repeat',
      willChange: 'transform'
    }, _extends2[horizontal ? 'height' : 'width'] = '100%', _extends2[horizontal ? 'width' : 'height'] = ctrl.get('space'), _extends2.WebkitTransform = translate3d, _extends2.MsTransform = translate3d, _extends2.transform = translate3d, _extends2), rest.style)
  }));
  return null;
});
var Parallax = React__default.memo(function (_ref2) {
  var _extends3;

  var pages = _ref2.pages,
      _ref2$config = _ref2.config,
      config = _ref2$config === void 0 ? core.config.slow : _ref2$config,
      _ref2$enabled = _ref2.enabled,
      enabled = _ref2$enabled === void 0 ? true : _ref2$enabled,
      _ref2$horizontal = _ref2.horizontal,
      horizontal = _ref2$horizontal === void 0 ? false : _ref2$horizontal,
      innerStyle = _ref2.innerStyle,
      rest = _objectWithoutPropertiesLoose(_ref2, ["pages", "config", "enabled", "horizontal", "innerStyle"]);

  var _useState = React.useState(false),
      ready = _useState[0],
      setReady = _useState[1];

  var state;
  state = useMemoOne.useMemoOne(function () {
    return {
      config: config,
      busy: false,
      space: 0,
      current: 0,
      offset: 0,
      controller: new core.Controller({
        scroll: 0
      }),
      layers: new Set(),
      update: function update() {
        return _update();
      },
      scrollTo: function scrollTo(offset) {
        return _scrollTo(offset);
      },
      stop: function stop() {
        return state.controller.stop();
      }
    };
  }, []);
  React.useEffect(function () {
    state.config = config;
  }, [config]);
  var containerRef = React.useRef();
  var contentRef = React.useRef();

  var _update = function _update() {
    var container = containerRef.current;
    if (!container) return;
    var spaceProp = horizontal ? 'clientWidth' : 'clientHeight';
    state.space = container[spaceProp];
    var scrollType = getScrollType(horizontal);

    if (enabled) {
      state.current = container[scrollType];
    } else {
      container[scrollType] = state.current = state.offset * state.space;
    }

    var content = contentRef.current;

    if (content) {
      var sizeProp = horizontal ? 'width' : 'height';
      content.style[sizeProp] = state.space * pages + "px";
    }

    state.layers.forEach(function (layer) {
      layer.setHeight(state.space, true);
      layer.setPosition(state.space, state.current, true);
    });
  };

  var _scrollTo = function _scrollTo(offset) {
    var container = containerRef.current;
    var scrollType = getScrollType(horizontal);
    state.offset = offset;
    state.controller.stop().start({
      scroll: offset * state.space,
      config: config,
      onFrame: function onFrame(_ref3) {
        var scroll = _ref3.scroll;
        container[scrollType] = scroll;
      }
    });
  };

  var onScroll = function onScroll(event) {
    if (!state.busy) {
      state.busy = true;
      state.current = event.target[getScrollType(horizontal)];
      globals.frameLoop.onFrame(function () {
        state.layers.forEach(function (layer) {
          return layer.setPosition(state.space, state.current);
        });
        state.busy = false;
      });
    }
  };

  React.useEffect(function () {
    return state.update();
  });
  shared.useOnce(function () {
    setReady(true);

    var onResize = function onResize() {
      var update = function update() {
        return state.update();
      };

      globals.frameLoop.onFrame(update);
      setTimeout(update, 150); // Some browsers don't fire on maximize!
    };

    window.addEventListener('resize', onResize, false);
    return function () {
      return window.removeEventListener('resize', onResize, false);
    };
  });
  var overflow = enabled ? 'scroll' : 'hidden';
  return React__default.createElement(globals.defaultElement, Object.assign({}, rest, {
    ref: containerRef,
    onScroll: onScroll,
    onWheel: enabled ? state.stop : null,
    onTouchStart: enabled ? state.stop : null,
    style: _extends({
      position: 'absolute',
      width: '100%',
      height: '100%',
      overflow: overflow,
      overflowY: horizontal ? 'hidden' : overflow,
      overflowX: horizontal ? overflow : 'hidden',
      WebkitOverflowScrolling: 'touch',
      WebkitTransform: START_TRANSLATE,
      MsTransform: START_TRANSLATE,
      transform: START_TRANSLATE_3D
    }, rest.style)
  }), ready && React__default.createElement(globals.defaultElement, {
    ref: contentRef,
    style: _extends((_extends3 = {
      overflow: 'hidden',
      position: 'absolute'
    }, _extends3[horizontal ? 'height' : 'width'] = '100%', _extends3[horizontal ? 'width' : 'height'] = state.space * pages, _extends3.WebkitTransform = START_TRANSLATE, _extends3.MsTransform = START_TRANSLATE, _extends3.transform = START_TRANSLATE_3D, _extends3), innerStyle)
  }, React__default.createElement(ParentContext.Provider, {
    value: state
  }, rest.children)));
});

exports.Parallax = Parallax;
exports.ParallaxLayer = ParallaxLayer;
